txp/Pos/StateLock/PrioLock.hs

```diff
@@ -0,0 +1,77 @@
+{-|
+Module:      Pos.StateLock.PrioLock
+Description: Provides a prioritised lock
+
+Provides a lock that can be taken with either high or low precedence.
+Within each precedence, the lock is taken in FIFO order.
+-}
+
+{-# LANGUAGE LambdaCase #-}
+
+module Pos.StateLock.PrioLock where
+
+import           Control.Concurrent.STM      (TMVar, newEmptyTMVar
+                                             , putTMVar, takeTMVar)
+import           Control.Monad.Catch         (MonadMask)
+import           Universum
+
+newtype PrioLock = PrioLock (TVar PrioLockState)
+
+data PrioLockState =
+    PUnlocked
+    | PLocked [TMVar ()] [TMVar ()] -- TODO: consider using queues instead of lists
+    -- ^ locked, with a list of contenders with high precedence, and a
+    -- second list with contenders of low precedence
+
+data Priority = High | Low
```

Comment:
We'll have to use these in other modules so I'd suggest bigger names for use in other contexts:

    data Priority = HighPriority | LowPriority